app/frontend/components/tags/tag_statistics.js

```diff
@@ -58,80 +97,79 @@ class TagStatistics {
       return;
     }
 
-    var overall_verse_count = 0;
+    var overallVerseCount = 0;
     for (let i = 0; i < chapterCount; i++) {
       let currentChapterVerseCount = allChapterVerseCounts[i];
-      overall_verse_count += currentChapterVerseCount;
+      overallVerseCount += currentChapterVerseCount;
     }
 
-    var tag_statistics_html = "<table class='tag-statistics'>";
-    tag_statistics_html += "<tr><th style='text-align: left;'>Tag</th>"
+    var tagStatisticsHTML = "<table class='tag-statistics'>";
+    tagStatisticsHTML += "<tr><th style='text-align: left;'>Tag</th>"
                         +  "<th style='text-align: left; width: 2em;'>#</th>"
                         +  "<th style='text-align: left; width: 2em;'>%</th></tr>";
     
-    const MAX_CLUSTERS = 6;
+    var [clusters, maxClusterPercentage] = this.getHighFrequencyClusters(overallVerseCount);
     var MIN_CLUSTER_PERCENT = 3;
 
-    var clusters = [];
-    var maxClusterPercentage = 0;
-
-    for (let i = 0; i < tags_by_verse_count.length; i++) {
-      let tag_title = tags_by_verse_count[i];
-      let tagged_verse_count = book_tag_statistics[tag_title];
-      let tagged_verse_percent = Math.round((tagged_verse_count / overall_verse_count) * 100);
-      
-      if (!clusters.includes(tagged_verse_percent) && clusters.length < MAX_CLUSTERS) {
-        if (tagged_verse_percent > maxClusterPercentage) {
-          maxClusterPercentage = tagged_verse_percent;
-        }
-
-        clusters.push(tagged_verse_percent);
-      }
-
-      if (clusters.length >= MAX_CLUSTERS) {
-        break;
-      }
-    }
-
     if (maxClusterPercentage == MIN_CLUSTER_PERCENT) {
       MIN_CLUSTER_PERCENT -= 1;
     }
 
     var wasMoreFrequent = true;
 
-    for (let i = 0; i < tags_by_verse_count.length; i++) {
-      let tag_title = tags_by_verse_count[i];
-      let tagged_verse_count = book_tag_statistics[tag_title];
-      if (tagged_verse_count == 0) {
+    for (let i = 0; i < this._tagsByVerseCount.length; i++) {
+      let tag_title = this._tagsByVerseCount[i];
+      let taggedVerseCount = this._currentBookTagStatistics[tag_title];
+      if (taggedVerseCount == 0) {
         continue;
       }
 
-      let tagged_verse_percent = Math.round((tagged_verse_count / overall_verse_count) * 100);
-      let current_row_html = "";
-
-      let isLessFrequent = tagged_verse_percent < MIN_CLUSTER_PERCENT || !clusters.includes(tagged_verse_percent);
+      let taggedVersePercent = Math.round((taggedVerseCount / overallVerseCount) * 100);
```

Comment:
I think it would be better to have a percentage from the overall amount of tags in the book (or even in all the books).
If I have only 10-20 different tags, I want the tag with 3 occurrences to be highlighted.

Reply:
You mean from the overall amount of (used) tags or overall amount of tagged verses / tag assignments?

Reply:
Overal amount of used tags. I think the idea that the number of occurances would be prety stable through out the different books and would vary only from overal tag amoun. I.e. even if I have just a few tags, most common tags should still be highlighted.

Everything else is ok.

Reply:
I'll implement this changed variant (changing the divisor in the formula) to see how it behaves.
Since tags / topics may vary greatly across books I would still suggest to keep it book-specific.

overallTaggedVerseCount should be easy to retrieve via [`ipcDb.getBookVerseTags`](https://github.com/ezra-bible-app/ezra-bible-app/blob/master/app/frontend/ipc/ipc_db.js#L99). I could use that instead of overallVerseCount. However, I have to filter out double verses.
But I think that's not what you indicate in your answer. What you mean is the number of tag assignments (one tag vs all tags), right?

How exactly would you define the "overall amount of used tags"?
This is easy to misinterpret ... (requirements :D)

Reply:
I think maybe just use `taggedVerseCount * 0.1` for 10%. I think it should be simple and transparent for the end-user.
My initial thought was to have just a number of tag occurrences > NUMBER_OF_FREQUENT_TAGS. But then the problem is that most of your tags would be highlighted as you have a lot of them. But I think this is the simplest option for the end user to understand why it's highlighted (in bold).

Reply:
The way how I adapted it now is this:

```
  getTagPercentage(bookTagStatistics, tagTitle, overallTagCount) {
    let taggedVerseCount = bookTagStatistics[tagTitle];
    let taggedVersePercent = Math.round((taggedVerseCount / overallTagCount) * 100);
    return taggedVersePercent;
  }
```

This in combination with `MIN_CLUSTER_PERCENT` (2) and `MAX_CLUSTERS` (4) then determines, which tags are most frequently used. It works well both with small collections of tags and big collections of tags.

Regarding the highlighting I think eventually we need to integrate some online help / hints into the app to help the user understand.

For now, I would like to keep it like this. Of course we can keep iterating based on personal usage experience and additional ideas!